agent/service.go

```diff
@@ -113,19 +115,42 @@ func unmarshalExtraData(marshalled *types.Any) (*proto.ExtraData, error) {
 	return extraData, nil
 }
 
-func fifoName(taskID, execID string) string {
-	return fmt.Sprintf("exec-%s-task-%s", execID, taskID)
+func (ts *TaskService) addCleanup(taskID, execID string, cleanup func() error) {
+	id := taskExecID(taskID, execID)
+	ts.execCleanups[id] = append(ts.execCleanups[id], cleanup)
+}
+
+func (ts *TaskService) doCleanup(taskID, execID string) error {
+	id := taskExecID(taskID, execID)
+
+	var err *multierror.Error
+	// iterate in reverse order so changes are "unwound" (similar to a defer)
+	for i := len(ts.execCleanups[id]) - 1; i >= 0; i-- {
+		err = multierror.Append(err, ts.execCleanups[id][i]())
+	}
+
+	delete(ts.execCleanups, id)
+	return err.ErrorOrNil()
 }
 
 // Create creates a new initial process and container using runc
-func (ts *TaskService) Create(requestCtx context.Context, req *taskAPI.CreateTaskRequest) (*taskAPI.CreateTaskResponse, error) {
+func (ts *TaskService) Create(requestCtx context.Context, req *taskAPI.CreateTaskRequest) (_ *taskAPI.CreateTaskResponse, err error) {
 	defer logPanicAndDie(log.G(requestCtx))
 	taskID := req.ID
 	execID := "" // the exec ID of the initial process in a task is an empty string by containerd convention
 
 	logger := log.G(requestCtx).WithField("TaskID", taskID).WithField("ExecID", execID)
 	logger.Info("create")
 
+	defer func() {
+		if err != nil {
+			cleanupErr := ts.doCleanup(taskID, execID)
+			if cleanupErr != nil {
+				logger.WithError(err).Error("failed to cleanup task")
```

Comment:
Do we not want to log the cleanup error as well?

Reply:
We do, this was an oversight, will fix